File: ios/SpaceXTransport.swift
//
//  SpaceXTransportScheduler.swift
//  openland
//
//  Created by Steve Kite on 5/13/19.
//  Copyright © 2019 Openland. All rights reserved.
//

import Foundation
import SwiftyJSON

enum SpaceXTransportOperationResult {
  case result(data: JSON)
  case error(error: JSON)
}

class SpaceXTransport {
  private let queue = DispatchQueue(label: "spacex-transport")
  private let transport: TransportState
  var connected: Bool = false
  var onConnected: (() -> Void)?
  var onDisconnected: (() -> Void)?
  
  init(url: String, params: [String: String?]) {
    NSLog("[SpaceX-Alloc]: init SpaceXTransportScheduler")
    self.transport = TransportState(url: url, params: params)
    self.transport.connectionCallback = { connected in
      self.queue.async {
        if self.connected != connected {
          self.connected = connected
          if connected {
            self.onConnected?()
          } else {
            self.onDisconnected?()
          }
        }
      }
    }
  }
  
  deinit {
    NSLog("[SpaceX-Alloc]: deinit SpaceXTransportScheduler")
  }
  
  func operation(
    operation: OperationDefinition,
    variables: JSON,
    queue: DispatchQueue,
    handler: @escaping  (SpaceXTransportOperationResult) -> Void
  ) {
    var isCompleted = false
    let _ = self.transport.operation(operation: operation, variables: variables) { (res) in
      self.queue.async {
        switch(res) {
        case .result(let data):
          if !isCompleted {
            isCompleted = true
            queue.async {
              handler(SpaceXTransportOperationResult.result(data: data))
            }
          }
        case .error(let data):
          if !isCompleted {
            isCompleted = true
            queue.async {
              handler(SpaceXTransportOperationResult.error(error: data))
            }
          }
        case .completed:
          // Ignore
          break
        }
      }
    }
  }
  
  func subscription(
    operation: OperationDefinition,
    variables: JSON,
    queue: DispatchQueue,
    handler: @escaping (SpaceXTransportOperationResult) -> Void
  ) -> RunningOperation {
    var isCompleted = false
    return self.transport.operation(operation: operation, variables: variables) { res in
      self.queue.async {
        switch(res) {
        case .result(let data):
          queue.async {
            if !isCompleted {
              handler(SpaceXTransportOperationResult.result(data: data))
            }
          }
        case .error(let error):
          queue.async {
            if !isCompleted {
              isCompleted = true
              handler(SpaceXTransportOperationResult.error(error: error))
            }
          }
        case .completed:
          queue.async {
            if !isCompleted {
              isCompleted = true
              handler(SpaceXTransportOperationResult.error(error: JSON([])))
            }
          }
        }
      }
    }
  }
  
  func close() {
    self.transport.connectionCallback = nil
    self.transport.close()
  }
}

File: android/app/src/main/java/com/openland/spacex/transport/SpaceXTransport.kt
package com.openland.spacex.transport

import com.openland.spacex.OperationDefinition
import com.openland.spacex.utils.DispatchQueue
import org.json.JSONArray
import org.json.JSONObject


sealed class TransportOperationResult {
    class Value(val data: JSONObject) : TransportOperationResult()
    class Error(val error: JSONArray) : TransportOperationResult()
}

class SpaceXTransport(val transport: TransportState) {
    fun operation(operation: OperationDefinition, variables: JSONObject, queue: DispatchQueue, callback: (result: TransportOperationResult) -> Unit) {
        var completed = false
        transport.operation(operation, variables) { res ->
            if (res is TransportResult.Result) {
                queue.async {
                    if (completed) {
                        return@async
                    }
                    completed = true
                    callback(TransportOperationResult.Value(res.data))
                }
            } else if (res is TransportResult.Error) {
                queue.async {
                    if (completed) {
                        return@async
                    }
                    completed = true
                    callback(TransportOperationResult.Error(res.error))
                }
            }
        }
    }

    fun subscription(operation: OperationDefinition, variables: JSONObject, queue: DispatchQueue, callback: (result: TransportOperationResult) -> Unit): () -> Unit {
        var completed = false
        return transport.operation(operation, variables) { res ->
            if (res is TransportResult.Result) {
                queue.async {
                    if (completed) {
                        return@async
                    }
                    callback(TransportOperationResult.Value(res.data))
                }
            } else if (res is TransportResult.Error) {
                queue.async {
                    if (completed) {
                        return@async
                    }
                    completed = true
                    callback(TransportOperationResult.Error(res.error))
                }
            } else if (res == TransportResult.Completed) {
                queue.async {
                    if (completed) {
                        return@async
                    }
                    completed = true
                    callback(TransportOperationResult.Error(JSONArray()))
                }
            }
        }
    }
}